gio/src/subclass/initable.rs

```diff
@@ -0,0 +1,215 @@
+        iface.init = Some(initable_init::<T>);
+    }
+
+    fn instance_init(_instance: &mut glib::subclass::InitializingObject<T>) {}
+}
+
+unsafe extern "C" fn initable_init<T: InitableImpl>(
+    initable: *mut ffi::GInitable,
+    cancellable: *mut ffi::GCancellable,
+    error: *mut *mut glib::ffi::GError,
+) -> glib::ffi::gboolean {
+    let instance = &*(initable as *mut T::Instance);
+    let imp = instance.impl_();
+
+    match imp.init(
+        from_glib_borrow::<_, Initable>(initable).unsafe_cast_ref(),
+        Option::<Cancellable>::from_glib_borrow(cancellable)
+            .as_ref()
+            .as_ref(),
+    ) {
+        Ok(()) => glib::ffi::GTRUE,
+        Err(e) => {
+            if !error.is_null() {
+                *error = e.into_raw();
+            }
+            glib::ffi::GFALSE
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::prelude::*;
+    use crate::traits::InitableExt;
+    use crate::{Cancellable, Initable};
+
+    pub mod imp {
+        use super::*;
+        use crate::Cancellable;
+        use crate::Initable;
+        use std::cell::Cell;
+
+        pub struct InitableTestType(pub Cell<u64>);
+
+        #[glib::object_subclass]
+        impl ObjectSubclass for InitableTestType {
+            const NAME: &'static str = "InitableTestType";
+            type Type = super::InitableTestType;
+            type ParentType = glib::Object;
+            type Interfaces = (Initable,);
+
+            fn new() -> Self {
+                Self(Cell::new(0))
+            }
+        }
+
+        impl InitableImpl for InitableTestType {
+            fn init(
+                &self,
+                initable: &Self::Type,
+                _cancellable: Option<&Cancellable>,
+            ) -> Result<(), glib::Error> {
+                InitableTestType::from_instance(initable)
+                    .0
+                    .set(0x123456789abcdef);
+                Ok(())
+            }
+        }
+
+        impl ObjectImpl for InitableTestType {}
+    }
+
+    pub mod ffi {
+        use super::*;
+        pub type InitableTestType = <imp::InitableTestType as ObjectSubclass>::Instance;
+
+        #[no_mangle]
+        pub unsafe extern "C" fn initable_test_type_get_type() -> glib::ffi::GType {
+            imp::InitableTestType::type_().into_glib()
+        }
+
+        #[no_mangle]
+        pub unsafe extern "C" fn initable_test_type_get_value(this: *mut InitableTestType) -> u64 {
+            let this = super::InitableTestType::from_glib_borrow(this);
+            imp::InitableTestType::from_instance(&this).0.get()
+        }
+    }
+
+    glib::wrapper! {
+        pub struct InitableTestType(ObjectSubclass<imp::InitableTestType>)
+            @implements Initable;
+    }
+
+    impl InitableTestType {
+        #[allow(clippy::new_without_default)]
+        pub fn new() -> Self {
+            glib::Object::new(&[]).expect("Failed creation of InitableTestType object")
```

Comment:
This should go via the initable interface

Reply:
I use the type in a couple of tests (`test_initable_with_init` and `test_initable_through_ffi`) which need the object in a non-initialized state. I can remove the tests, but I think the FFI one is good to check that a declaration of a Rust initable type would work called from C

Reply:
`test_initable_with_initable_new` does the test using the initable interface

Reply:
Ok then, but please add a comment about this :)